bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -0,0 +1,265 @@
+where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	XcmPrologueFee: Get<u128>,
+{
+	fn convert(
+		message: Message,
+		origin_account_location: Location,
+	) -> Result<Xcm<()>, ConvertMessageError> {
+		let mut message_xcm: Xcm<()> = Xcm::new();
+		if message.xcm.len() > 0 {
+			// Decode xcm
+			let versioned_xcm = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			)
+			.map_err(|_| ConvertMessageError::InvalidVersionedXCM)?;
+			message_xcm = versioned_xcm.try_into().map_err(|_| ConvertMessageError::InvalidXCM)?;
+		}
+
+		log::debug!(target: LOG_TARGET,"xcm decoded as {:?}", message_xcm);
+
+		let network = EthereumNetwork::get();
+
+		let origin_location = Location::new(
+			2,
+			[GlobalConsensus(network), AccountKey20 { key: message.origin.into(), network: None }],
+		);
+
+		let network = EthereumNetwork::get();
+
+		let fee_asset = Location::new(1, Here);
+		let fee: xcm::prelude::Asset = (fee_asset.clone(), XcmPrologueFee::get()).into();
+		let mut instructions = vec![
+			ReceiveTeleportedAsset(fee.clone().into()),
+			PayFees { asset: fee },
+			DescendOrigin(PalletInstance(InboundQueuePalletInstance::get()).into()),
+			UniversalOrigin(GlobalConsensus(network)),
+		];
+
+		for asset in &message.assets {
+			match asset {
+				Asset::NativeTokenERC20 { token_id, value } => {
+					let token_location: Location = Location::new(
+						2,
+						[
+							GlobalConsensus(EthereumNetwork::get()),
+							AccountKey20 { network: None, key: (*token_id).into() },
+						],
+					);
+					instructions.push(ReserveAssetDeposited((token_location, *value).into()));
+				},
+				Asset::ForeignTokenERC20 { token_id, value } => {
+					let asset_id = ConvertAssetId::convert(&token_id)
+						.ok_or(ConvertMessageError::InvalidAsset)?;
+					instructions.push(WithdrawAsset((asset_id, *value).into()));
+				},
+			}
+		}
+
+		if let Some(claimer) = message.claimer {
+			let claimer = Junction::decode(&mut claimer.as_ref())
+				.map_err(|_| ConvertMessageError::InvalidClaimer)?;
+			let claimer_location: Location = Location::new(0, [claimer.into()]);
+			instructions.push(SetAssetClaimer { location: claimer_location });
+		}
+
+		// Set the alias origin to the original sender on Ethereum. Important to be before the
+		// arbitrary XCM that is appended to the message on the next line.
+		instructions.push(AliasOrigin(origin_location.into()));
+
+		// Add the XCM sent in the message to the end of the xcm instruction
+		instructions.extend(message_xcm.0);
+
+		let appendix = vec![
+			RefundSurplus,
+			// Refund excess fees to the relayer
+			// TODO maybe refund all fees to the relayer instead of just DOT?
+			DepositAsset {
+				assets: Wild(AllOf { id: AssetId(fee_asset.into()), fun: WildFungible }),
+				beneficiary: origin_account_location,
+			},
```

Comment:
@vgeddes @yrong please let me know what you think of this? The user may specify other fee assets in the arbitrary XCM, right? So perhaps this should be `AllCounted`?

Reply:
If the extra fee(in WETH) is refunded to the relayer, we may not need to deposit it into RewardLeger?

Reply:
Hmmm are they not separate though? Ie the reward is specified here: https://github.com/claravanstaden/polkadot-sdk/pull/4/files#diff-51b54cea1a87bc694f384034a09cd94281e472795011f662bdb106deb9db2699R23 For example, say the fee is 1 WETH and the execution fee is 0.2 WETH, fee provided was 0.3 WETH. So 0.1 WETH will be refunded but the relayer should actually received the full 1 WETH?

Reply:
> say the fee is 1 WETH, ... fee provided was 0.3 WETH

I'd assume fee provided should be same as the Ether value from envelop which is already charged from user. 

https://github.com/Snowfork/snowbridge/blob/553b0b9d9ec6f3683c6c605417ea5532282771ed/contracts/src/v2/Calls.sol#L110 

Then we don't need a separated reward process.